### DISCUSSION TITLE
Open in DAX studio button

### DISCUSSION THREAD

**Discussion by bernatagulloesbrina** [TabularEditor/TabularEditor3#1067] [Ideas]
Sometimes you have a nice dax query written with all the intellisense, but want to have a look at server timings 
wouldn't it be great if we could just click a "open in dax studio" button? way cooler than copy paste, that's for sure. Super critical? probably not.

**Comment by otykier** [comment]
It's a good idea. In fact, you can do it already today with a (quite ugly) c# script:

**DISCLAIMER: This script uses an undocumented API + reflection. It is not guaranteed to work in future versions of TE3**

```csharp
using TabularEditor.Shared;
using System.IO;
using System.Diagnostics;
using System.Reflection;

// Get the current document:
var document = ((dynamic)SharedApp.Instance).ViewService.CurrentDocument;
if(document == null) return;

// Get the code within the current document:
var code = document.GetType()
    .GetProperty("TabularEditor.UI.Views.IWorkDocumentView.Content", BindingFlags.Instance | BindingFlags.NonPublic)
    .GetValue(document);

// Get a (temporary) file name:
var file = Path.ChangeExtension(Path.Combine(Path.GetTempPath(), document.Caption.ToString()), ".dax");

// Save the file:
File.WriteAllText(file, code);

// Launch DAX Studio with the -f argument to load the file
// TODO: Make sure the installation path of DAX studio matches the below
Process.Start("c:\\Program Files\\DAX Studio\\DaxStudio.exe", "-f \"" + file + "\"");
```

Save this script as a macro (set the macro context to "Model"):

![image](https://user-images.githubusercontent.com/8976200/136763325-40be2bb6-afd0-41fe-9c5d-c9badd305032.png)

Then use the **Tools > Customize** feature to add this macro to the DAX Query toolbar:

![image](https://user-images.githubusercontent.com/8976200/136763408-bcf20fc2-683b-4aca-bea4-2b57082fe1a1.png)

![image](https://user-images.githubusercontent.com/8976200/136763472-0c091fe4-95a8-49f5-8303-d92b6846941a.png)

Result:

![launch-in-daxstudio](https://user-images.githubusercontent.com/8976200/136763835-e2874d14-c851-43ea-99dd-b8dc07cae69a.gif)

**Comment by otykier** [comment]
Here's an improved version of the script, that also instructs DAX Studio to connect to Analysis Services:

```csharp
#r "Microsoft.AnalysisServices.Core.dll"

using TabularEditor.Shared;
using System.IO;
using System.Diagnostics;
using System.Reflection;

// Get the current document:
var document = ((dynamic)SharedApp.Instance).ViewService.CurrentDocument;
if(document == null) return;

// Get the code within the current document:
var code = document.GetType()
    .GetProperty("TabularEditor.UI.Views.IWorkDocumentView.Content", BindingFlags.Instance | BindingFlags.NonPublic)
    .GetValue(document);

// Get a (temporary) file name:
var file = Path.ChangeExtension(Path.Combine(Path.GetTempPath(), document.Caption.ToString()), ".dax");

// Get Analysis Services connection details:
var connectionInfo = Model.Database.TOMDatabase.Server.ConnectionInfo;
var server = connectionInfo.Server + ":" + connectionInfo.Port;
var database = Model.Handler.OriginalDatabaseId;

// Save the file:
File.WriteAllText(file, code);

// Launch DAX Studio with the -f argument to load the file, and the -s and -d arguments to connect to Analysis Services:
Process.Start("c:\\Program Files\\DAX Studio\\DaxStudio.exe", "-s \"" + server + "\" -d \"" + database + "\" -f \"" + file + "\"");
```

**Comment by bernatagulloesbrina** [comment]
OMG!
That's some next level scripting!!

I'll look at it in detail

Regards and thank you!

**Comment by samaguire** [comment]
Nice one guys! 😎

**Comment by samaguire** [comment]
I've just tweaked the code to work with both Power BI Desktop and Power BI Premium connections

``` csharp
#r "Microsoft.AnalysisServices.Core.dll"

using TabularEditor.Shared;
using System.IO;
using System.Diagnostics;
using System.Reflection;

// Get the current document:
var document = ((dynamic)SharedApp.Instance).ViewService.CurrentDocument;
if(document == null) return;

// Get the code within the current document:
var code = document.GetType()
    .GetProperty("TabularEditor.UI.Views.IWorkDocumentView.Content", BindingFlags.Instance | BindingFlags.NonPublic)
    .GetValue(document);

// Get a (temporary) file name:
var file = Path.ChangeExtension(Path.Combine(Path.GetTempPath(), document.Caption.ToString()), ".dax");

// Get Analysis Services connection details:
var connectionInfo = Model.Database.TOMDatabase.Server.ConnectionInfo;
var server = "";
var database = "";
if (connectionInfo.Port == null)
{
    server = connectionInfo.Server;
    database = Model.Database.Name;
}
else
{
    server = connectionInfo.Server + ":" + connectionInfo.Port;
    database = Model.Database.ID;
}

// Save the file:
File.WriteAllText(file, code);

// Launch DAX Studio with the -f argument to load the file, and the -s and -d arguments to connect to Analysis Services:
Process.Start("c:\\Program Files\\DAX Studio\\DaxStudio.exe", "-s \"" + server + "\" -d \"" + database + "\" -f \"" + file + "\"");
```

**Comment by pawarbi** [comment]
Hi @otykier , I running into an error. I followed the steps. Copied the C# Script, saved it as a Macro. When I run the Macro, I am getting below error:
Any idea what could be causing it? I am on TE 3.2.3
![image](https://user-images.githubusercontent.com/62612119/189447624-313ab081-aa92-47be-a4d6-a212d21eadd3.png)

**Comment by otykier** [comment]
Hi @pawarbi 
Unfortunately, the script uses an undocumented API + reflection, which should generally be avoided. I should have put a disclaimer when I posted the script initially, as this technique is subject to breaking when we change internal stuff. We will make sure to provide a fully supported/documented way to access TE3 document content in future versions.
